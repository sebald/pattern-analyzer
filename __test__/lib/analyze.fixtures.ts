export const empire = [
  [
    'backstabber-battleofyavin',
    'blacksquadronace',
    'darthvader-battleofyavin',
    'flightleaderubbel',
    'maulermithel-battleofyavin',
  ],
  [
    'backstabber-battleofyavin',
    'captainferoph',
    'darthvader-battleofyavin',
    'maulermithel-battleofyavin',
    'moffgideon',
  ],
  [
    'blacksquadronace',
    'blacksquadronace',
    'darthvader-tieddefender',
    'idenversio',
    'vizier',
  ],
  ['darthvader-battleofyavin', 'junoeclipse', 'maarekstele', 'majorvermeil'],
  [
    'darthvader-battleofyavin',
    'idenversio',
    'moffgideon',
    'rearadmiralchiraneau',
  ],
  [
    'backstabber-battleofyavin',
    'blacksquadronace',
    'darthvader-battleofyavin',
    'flightleaderubbel',
    'maulermithel-battleofyavin',
  ],
  [
    'backstabber-battleofyavin',
    'blacksquadronace',
    'darthvader-battleofyavin',
    'flightleaderubbel',
    'maulermithel-battleofyavin',
  ],
  ['captainferoph', 'darthvader-tieddefender', 'idenversio', 'vizier'],
  ['captainferoph', 'darthvader-tieddefender', 'idenversio', 'vizier'],
  [
    'backstabber-battleofyavin',
    'darthvader-battleofyavin',
    'idenversio',
    'maulermithel-battleofyavin',
    'vizier',
  ],
  [
    'backstabber-battleofyavin',
    'darthvader-battleofyavin',
    'idenversio',
    'maulermithel-battleofyavin',
    'vizier',
  ],
  [
    'blacksquadronace',
    'darthvader-tieddefender',
    'gideonhask',
    'maulermithel',
    'wampa',
  ],
  [
    'blacksquadronace',
    'blacksquadronace',
    'darthvader-battleofyavin',
    'idenversio',
    'mornakee',
  ],
  [
    'blacksquadronace',
    'blacksquadronace',
    'darthvader-tieddefender',
    'idenversio',
    'vizier',
  ],
  [
    'backstabber-battleofyavin',
    'blacksquadronace',
    'blacksquadronace',
    'darthvader-battleofyavin',
    'idenversio',
    'maulermithel-battleofyavin',
  ],
  [
    'backstabber-battleofyavin',
    'darthvader-battleofyavin',
    'idenversio',
    'junoeclipse',
    'maulermithel-battleofyavin',
  ],
  ['darthvader-tieddefender', 'moffgideon', 'puresabacc', 'vizier'],
  [
    'captainferoph',
    'scarifbasepilot',
    'scarifbasepilot',
    'scarifbasepilot',
    'vizier',
  ],
  ['academypilot', 'captainferoph', 'darthvader-tieddefender', 'secondsister'],
  [
    'blacksquadronace',
    'countdown',
    'darthvader-battleofyavin',
    'flightleaderubbel',
    'maulermithel-battleofyavin',
  ],
  [
    'backstabber-battleofyavin',
    'darthvader-battleofyavin',
    'howlrunner',
    'idenversio',
    'maulermithel-battleofyavin',
  ],
];
