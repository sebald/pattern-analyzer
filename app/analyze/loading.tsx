import { Skeleton, HeadlineSkeleton, CardChartSkeleton } from '@/ui';

export const Loading = () => (
  <Skeleton>
    <HeadlineSkeleton className="mt-8 h-14 w-72" />
    <HeadlineSkeleton className="mt-2 h-4 w-44" />
    <div className="mb-4 mt-8 flex flex-col items-end">
      <HeadlineSkeleton className="h-8 w-44" />
    </div>
    <div className="grid grid-cols-2 gap-4">
      <CardChartSkeleton />
      <CardChartSkeleton />
      <CardChartSkeleton />
      <CardChartSkeleton />
    </div>
  </Skeleton>
);

export default Loading;
