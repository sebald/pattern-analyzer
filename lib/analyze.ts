import { createSubsets, isSubset, round } from './utils';

const SEPARATOR = '|';

export interface AnalyzeOotions {
  threshold?: number;
}

export interface AnalyzeResult {
  refs: number[];
  set: string[];
  occurrence: number;
  score: number;
}

export const analyze = (
  squads: string[][],
  { threshold = 2 }: AnalyzeOotions = {}
) => {
  const total = squads.length;
  let shipCount = 0; // used to calculate weighted average

  let subsets = new Set<string>();

  squads.forEach(squad => {
    shipCount = shipCount + squad.length;

    const sets = createSubsets(squad).map(set => {
      set.sort();
      return set.join(SEPARATOR);
    });

    subsets = new Set([...subsets, ...sets]);
  });

  // Calculate weighted averge
  const shipAverage = round(shipCount / total, 2);

  /**
   * Store subset -> ids of squad it is included,
   * where id is the index in the squads array.
   */
  const map = new Map<string, number[]>();

  subsets.forEach(key => {
    const set = key.split(SEPARATOR);
    squads.forEach((squad, idx) => {
      if (isSubset(set, squad)) {
        const entry = map.get(key) || [];
        entry.push(idx);
        map.set(key, entry);
      }
    });
  });

  let result: AnalyzeResult[] = [];

  map.forEach((entry, key) => {
    // Skip and remove if below threshold
    if (entry.length < threshold) {
      map.delete(key);
      return;
    }

    const set = key.split(SEPARATOR);
    if (set.length < 2) {
      return;
    }

    const occurrence = round(entry.length / total, 2);
    const score = round((set.length / shipAverage) * occurrence * 100, 2);

    result.push({
      refs: entry,
      set,
      occurrence,
      score,
    });
  });

  /**
   * Next:
   *
   * threshold for # of lists
   * threshold for size of combination/subset
   *
   * calculate ocrruance (appearance in lists / total squads)
   * weight: large combination = better
   *
   * subset weight = subset size / average # of ships +/* occurance percentage
   */

  // TODO: MAKE THRESHOLD FOR SET SIZE CONFIGURABLE

  return result;
};
