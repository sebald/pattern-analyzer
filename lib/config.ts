// Env
// ---------------
export const baseUrl =
  process.env.NEXT_PUBLIC_APP_URL ??
  `https://${process.env.NEXT_PUBLIC_VERCEL_URL}`;

// Settings
// ---------------
export const pointsUpdateDate = '2023-06-01';

export const vendors = [
  { id: 'listfortress', name: 'Listfortress' },
  { id: 'longshanks', name: 'Longshanks' },
  { id: 'rollbetter', name: 'Rollbetter' },
];

// Navigation
// ---------------
export const siteNavigation = [
  {
    name: 'Tournaments',
    href: '/tournament',
  },
  {
    name: 'Insights',
    href: '/insights',
  },
  {
    name: 'Compositions',
    href: '/composition',
  },
  {
    name: 'Pilots',
    href: '/pilot',
  },
  {
    name: 'Upgrades',
    href: '/upgrade',
  },
];

export const secondaryNavigation = [
  {
    name: 'About',
    href: '/about',
  },
  {
    name: 'GitHub',
    href: 'https://github.com/sebald/pattern-analyzer',
  },
];

export const xWingResources = [
  {
    name: 'Listfortess',
    href: 'https://listfortress.com/',
  },
  {
    name: 'MetaWing',
    href: 'https://meta.listfortress.com/',
  },
  {
    name: 'Pink Brain Matter',
    href: 'https://pinksquadron.dk/pbm/',
  },
  {
    name: 'Advanced Targeting Computer',
    href: 'http://advancedtargeting.computer/',
  },
];
