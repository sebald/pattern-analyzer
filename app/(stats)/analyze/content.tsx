import { Suspense } from 'react';
import { getSquads, getFactionCount } from '@/lib/db/squads';
import { getTournamentsCount } from '@/lib/db/tournaments';
import { ChassisDistribution } from '@/ui/stats/chassis-distribution';
import { CompositionStats } from '@/ui/stats/composition-stats';
import { FactionDistribution } from '@/ui/stats/faction-distribution';
import { FactionPerformance } from '@/ui/stats/faction-performance';
import { FactionVictories } from '@/ui/stats/faction-victories';
import { PilotCostDistribution } from '@/ui/stats/pilot-cost-distribution';
import { PilotSkillDistribution } from '@/ui/stats/pilot-skill-distribution';
import { CardChartSkeleton } from '@/ui/skeleton';
import { PilotStats } from '@/ui/stats/pilot-stats';
import { SquadSize } from '@/ui/stats/squad-size';
import { StatsHint } from '@/ui/stats/stats-hint';
import { UpgradeStats } from '@/ui/stats/upgrade-stats';
import { setup } from '@/lib/stats';
import {
  CompositionData,
  FactionData,
  PilotData,
  PilotCostDistributionData,
  PilotSkillDistributionData,
  ShipData,
  SquadSizeData,
  UpgradeData,
  composition,
  faction,
  pilotCostDistribution,
  pilotSkillDistribution,
  pilot,
  ship,
  squadSize,
  upgrade,
} from '@/lib/stats/module';

// Helpers
// ---------------
const Loading = () => (
  <div className="grid grid-cols-1 gap-4 md:grid-cols-2">
    <CardChartSkeleton />
    <CardChartSkeleton />
    <CardChartSkeleton />
    <CardChartSkeleton />
  </div>
);

interface StatsData
  extends CompositionData,
    FactionData,
    PilotData,
    PilotCostDistributionData,
    PilotSkillDistributionData,
    ShipData,
    SquadSizeData,
    UpgradeData {}

const create = setup<StatsData>([
  composition,
  faction,
  pilotCostDistribution,
  pilotSkillDistribution,
  pilot,
  ship,
  squadSize,
  upgrade,
]);

// Props
// ---------------
export interface ContentProps {
  from: Date;
  to?: Date;
}

// Component
// ---------------
const AsyncContent = async ({ from, to }: ContentProps) => {
  const [squads, tournaments, count] = await Promise.all([
    getSquads({ from, to }),
    getTournamentsCount({ from, to }),
    getFactionCount({ from, to }),
  ]);

  const stats = create(squads, {
    count,
    tournaments,
    smallSamples: false,
  });

  return (
    <>
      <div className="grid grid-cols-1 gap-4 md:grid-cols-12">
        <div className="md:col-span-6">
          <FactionDistribution
            value={stats.faction}
            total={count.all}
            ignoreUnknown
          />
        </div>
        <div className="md:col-span-6">
          <FactionPerformance value={stats.faction} ignoreUnknown />
        </div>
        <div className="md:col-span-6">
          <FactionVictories value={stats.faction} total={tournaments} />
        </div>
        <div className="md:col-span-6">
          <SquadSize
            value={stats.squadSize}
            total={count.all - count.unknown}
          />
        </div>
        <div className="col-span-full">
          <ChassisDistribution value={stats.ship} />
        </div>
        <div className="md:col-span-6">
          <PilotCostDistribution value={stats.pilotCostDistribution} />
        </div>
        <div className="md:col-span-6">
          <PilotSkillDistribution value={stats.pilotSkillDistribution} />
        </div>
        <div className="col-span-full">
          <PilotStats value={stats.pilot} />
        </div>
        <div className="col-span-full">
          <UpgradeStats value={stats.upgrade} />
        </div>
        <div className="col-span-full">
          <CompositionStats value={stats.composition} />
        </div>
        <div className="col-span-full pt-8 lg:col-start-2 lg:col-end-12">
          <StatsHint />
        </div>
      </div>
    </>
  );
};

export const Content = (props: ContentProps) => (
  <Suspense fallback={<Loading />}>
    <AsyncContent {...props} />
  </Suspense>
);
