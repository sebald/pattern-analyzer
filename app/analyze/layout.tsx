import { BASE_URL } from '@/lib/env';
import { Container, Link, Logo } from '@/ui';

// Metadata
// ---------------
export const metadata = {
  title: 'Pattern Analyzer | Analyze',
  description: 'Analyze the current X-Wing meta!',
  openGraph: {
    siteName: 'Pattern Analyzer',
    title: 'Analyze',
    description: 'Analyze the current X-Wing meta!',
    images: `${BASE_URL}/api/og.png?title`,
    locale: 'en-US',
    type: 'website',
  },
};

// Component
// ---------------
const Layout = ({ children }: { children: React.ReactNode }) => (
  <>
    <nav className="px-2 py-2 text-center leading-none xl:px-6">
      <Link className="inline-block" href="/">
        <Logo className="h-7 w-7 text-primary-400 hover:text-primary-900 lg:h-9 lg:w-9" />
      </Link>
    </nav>
    <main className="pt-5 lg:pt-5">
      <Container>{children}</Container>
    </main>
  </>
);

export default Layout;
