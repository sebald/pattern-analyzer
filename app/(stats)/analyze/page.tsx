import { createMetadata } from '@/lib/metadata';
import { toDateRange } from '@/lib/utils/params.utils';

import { Caption, Message, Title } from '@/ui';
import { Filter } from '@/ui/params/filter';
import { DateRangeFilter } from '@/ui/params/date-range-filter';
import { StatsInfo } from '@/ui/stats/stats-info';

import { Content } from './content';

// Metadata
// ---------------
export const metadata = createMetadata({
  title: 'Analyze',
  description: 'Analyze the current X-Wing meta!',
});

// Props
// ---------------
interface AnalyzePageProps {
  searchParams: {
    from: string;
    to: string;
    'small-samples': 'show' | 'hide';
  };
}

// Page
// ---------------
const AnalyzePage = async ({ searchParams }: AnalyzePageProps) => {
  const params = toDateRange(searchParams);

  if (params.error) {
    return (
      <div className="grid flex-1 place-items-center">
        <Message variant="error">
          <Message.Title>Whoopsie, something went wrong!</Message.Title>
          Looks like there is an error in the given query parameters.
        </Message>
      </div>
    );
  }

  const { from, to } = params;

  return (
    <>
      <div className="pb-6">
        <Title>Analyze</Title>
        <Caption>
          <StatsInfo from={from} to={to} />
        </Caption>
      </div>
      <Filter>
        <DateRangeFilter />
      </Filter>
      <Content from={from} to={to} />
    </>
  );
};

export default AnalyzePage;
