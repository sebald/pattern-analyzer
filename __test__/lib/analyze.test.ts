import { empire } from './analyze.fixtures';
import { analyze } from 'lib/analyze';

test('analyze', () => {
  expect(analyze(empire)).toMatchSnapshot();
});
