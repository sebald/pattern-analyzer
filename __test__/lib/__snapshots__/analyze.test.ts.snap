// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze 1`] = `
[
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      14,
    ],
    "score": 7.98,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
    ],
  },
  {
    "occurrence": 0.43,
    "refs": [
      0,
      1,
      5,
      6,
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 18.07,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
    ],
  },
  {
    "occurrence": 0.29,
    "refs": [
      0,
      5,
      6,
      12,
      14,
      19,
    ],
    "score": 12.18,
    "set": [
      "blacksquadronace",
      "darthvader-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      14,
    ],
    "score": 11.97,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "darthvader-battleofyavin",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 5.88,
    "set": [
      "backstabber-battleofyavin",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 7.98,
    "set": [
      "blacksquadronace",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 8.82,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 7.98,
    "set": [
      "darthvader-battleofyavin",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 8.82,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 11.97,
    "set": [
      "blacksquadronace",
      "darthvader-battleofyavin",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 11.76,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "darthvader-battleofyavin",
      "flightleaderubbel",
    ],
  },
  {
    "occurrence": 0.43,
    "refs": [
      0,
      1,
      5,
      6,
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 18.07,
    "set": [
      "backstabber-battleofyavin",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      0,
      5,
      6,
      14,
      19,
    ],
    "score": 10.08,
    "set": [
      "blacksquadronace",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      14,
    ],
    "score": 11.97,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.48,
    "refs": [
      0,
      1,
      5,
      6,
      9,
      10,
      14,
      15,
      19,
      20,
    ],
    "score": 20.17,
    "set": [
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.43,
    "refs": [
      0,
      1,
      5,
      6,
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 27.1,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      0,
      5,
      6,
      14,
      19,
    ],
    "score": 15.13,
    "set": [
      "blacksquadronace",
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      14,
    ],
    "score": 15.97,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 7.98,
    "set": [
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 8.82,
    "set": [
      "backstabber-battleofyavin",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 11.97,
    "set": [
      "blacksquadronace",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 11.76,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 11.97,
    "set": [
      "darthvader-battleofyavin",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 11.76,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      0,
      5,
      6,
      19,
    ],
    "score": 15.97,
    "set": [
      "blacksquadronace",
      "darthvader-battleofyavin",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      0,
      5,
      6,
    ],
    "score": 14.71,
    "set": [
      "backstabber-battleofyavin",
      "blacksquadronace",
      "darthvader-battleofyavin",
      "flightleaderubbel",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      1,
      4,
    ],
    "score": 4.2,
    "set": [
      "darthvader-battleofyavin",
      "moffgideon",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      2,
      12,
      13,
      14,
    ],
    "score": 7.98,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      2,
      11,
      13,
    ],
    "score": 5.88,
    "set": [
      "blacksquadronace",
      "darthvader-tieddefender",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-tieddefender",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      2,
      12,
      13,
      14,
    ],
    "score": 7.98,
    "set": [
      "blacksquadronace",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      2,
      12,
      13,
      14,
    ],
    "score": 11.97,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      2,
      7,
      8,
      13,
    ],
    "score": 7.98,
    "set": [
      "darthvader-tieddefender",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "darthvader-tieddefender",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 8.4,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-tieddefender",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 4.2,
    "set": [
      "blacksquadronace",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "vizier",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      2,
      7,
      8,
      13,
      16,
    ],
    "score": 10.08,
    "set": [
      "darthvader-tieddefender",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "darthvader-tieddefender",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 8.4,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-tieddefender",
      "vizier",
    ],
  },
  {
    "occurrence": 0.29,
    "refs": [
      2,
      7,
      8,
      9,
      10,
      13,
    ],
    "score": 12.18,
    "set": [
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 8.4,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.19,
    "refs": [
      2,
      7,
      8,
      13,
    ],
    "score": 11.97,
    "set": [
      "darthvader-tieddefender",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 8.4,
    "set": [
      "blacksquadronace",
      "darthvader-tieddefender",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      2,
      13,
    ],
    "score": 10.5,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-tieddefender",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      3,
      15,
    ],
    "score": 4.2,
    "set": [
      "darthvader-battleofyavin",
      "junoeclipse",
    ],
  },
  {
    "occurrence": 0.33,
    "refs": [
      4,
      9,
      10,
      12,
      14,
      15,
      20,
    ],
    "score": 13.87,
    "set": [
      "darthvader-battleofyavin",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      7,
      8,
      18,
    ],
    "score": 5.88,
    "set": [
      "captainferoph",
      "darthvader-tieddefender",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      7,
      8,
    ],
    "score": 4.2,
    "set": [
      "captainferoph",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      7,
      8,
    ],
    "score": 6.3,
    "set": [
      "captainferoph",
      "darthvader-tieddefender",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.14,
    "refs": [
      7,
      8,
      17,
    ],
    "score": 5.88,
    "set": [
      "captainferoph",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      7,
      8,
    ],
    "score": 6.3,
    "set": [
      "captainferoph",
      "darthvader-tieddefender",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      7,
      8,
    ],
    "score": 6.3,
    "set": [
      "captainferoph",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      7,
      8,
    ],
    "score": 8.4,
    "set": [
      "captainferoph",
      "darthvader-tieddefender",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 10.08,
    "set": [
      "backstabber-battleofyavin",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 15.13,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 10.08,
    "set": [
      "idenversio",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 15.13,
    "set": [
      "backstabber-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 15.13,
    "set": [
      "darthvader-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.24,
    "refs": [
      9,
      10,
      14,
      15,
      20,
    ],
    "score": 20.17,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 4.2,
    "set": [
      "backstabber-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 4.2,
    "set": [
      "darthvader-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "backstabber-battleofyavin",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "darthvader-battleofyavin",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 8.4,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "idenversio",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 4.2,
    "set": [
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "backstabber-battleofyavin",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 8.4,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 6.3,
    "set": [
      "idenversio",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 8.4,
    "set": [
      "backstabber-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 8.4,
    "set": [
      "darthvader-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      9,
      10,
    ],
    "score": 10.5,
    "set": [
      "backstabber-battleofyavin",
      "darthvader-battleofyavin",
      "idenversio",
      "maulermithel-battleofyavin",
      "vizier",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      12,
      14,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-battleofyavin",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      12,
      14,
    ],
    "score": 6.3,
    "set": [
      "blacksquadronace",
      "darthvader-battleofyavin",
      "idenversio",
    ],
  },
  {
    "occurrence": 0.1,
    "refs": [
      12,
      14,
    ],
    "score": 8.4,
    "set": [
      "blacksquadronace",
      "blacksquadronace",
      "darthvader-battleofyavin",
      "idenversio",
    ],
  },
]
`;
